### PR TITLE
Resolve pylint c-class: C0301,C3001,C0114,C0115,C0116,C0200,C0121,C0201,C0206,C0209,C0412,C0415,C2801

### PR DESCRIPTION
**Pull request by SamuelMarks**
# Description

Linting is fun?

# Tests

N/A

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: .github/workflows/CPUTests.yml [hunk 1]
```diff
@@ -33,7 +33,7 @@ jobs:
         pytype --jobs auto --disable 'import-error,late-directive,wrong-arg-types,module-attr,unsupported-operands' MaxText/ || true
     - name: Analysing the code with pylint in Maxtext/
       run: |
-         pylint --disable C0301,C3001,C0114,C0115,C0116,C0200,C0121,C0201,C0206,C0209,C0412,C0415,C2801,E0102,E0606,E1102,E1111,E1123,E1135,E1136,R0401,R1701,R1703,R1710,R1711,R1735,R0917,R1714,R1716,R1719,R1721,R1728,R1728,W0102,W0107,W0201,W0212,W0221,W0237,W0404,W0611,W0612,W0613,W0621,W0622,W0631,W0707,W0718,W1201,W1203,W1309,W1514,W4901 MaxText/ && \
+         pylint --disable C0301,C3001,C0114,C0115,C0116,C0200,C0121,C0201,C0206,C0209,C0412,C0415,C2801,E0102,E0606,E0611,E1102,E1111,E1120,E1121,E1123,E1135,E1136,R0401,R1701,R1703,R1710,R1711,R1735,R0917,R1714,R1716,R1719,R1721,R1728,R1728,W0102,W0107,W0201,W0212,W0221,W0223,W0237,W0404,W0611,W0612,W0613,W0621,W0622,W0631,W0707,W0718,W1201,W1203,W1309,W1514,W4901 MaxText/ && \
```
Comment (shralex) [inline .github/workflows/CPUTests.yml:36]: can we remove the errors this PR resolves ?

Comment (SamuelMarks) [reply .github/workflows/CPUTests.yml:36]: Yes, once I've resolved them all; hold on almost finished

**Review by shralex — COMMENTED** [review]
Thanks Samuel!

Two general comments. 
As much as possible, when fixing these lets adhere to the Google Python styleguide (https://g3doc.corp.google.com/eng/doc/devguide/py/style/index.md?cl=head), as we want to improve the quality of the code and will eventually use Google's lint, if we're not already. In particular here, the first sentence of doc-strings (just the ones you're adding or changing) should be capitalized (except when starting with a variable name), and have a dot at the end of a sentence. Thank you!!

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -332,6 +336,7 @@ def generate_attention_mask(
       previous_chunk: Any = None,
       bidirectional_mask: Any = None,
   ) -> Array | None:
+    """Generate attention mask following Pallas MHA Flash Attention Reference."""
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/layers/attentions.py:339]: @shralex sure thing. In fact, the plan was once all the docstrings are in place to run gemini across the whole codebase to improve all the docstrings.

File: MaxText/layers/llama4.py [hunk 3]
```diff
@@ -133,7 +133,7 @@ def __call__(
 
     assert (
         cfg.num_experts >= 1
-    ), "Expected the Llama4 config to have num_experts > 1.  Note that MaxText only supports Llama4 Scout (MoE-only) for now!"
+    ), "Expected the Llama4 config to have num_experts > 1.  Note: MaxText only supports Llama4 Scout (MoE-only)"
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/layers/llama4.py:136]: Ok

File: MaxText/layers/moe.py [hunk 4]
```diff
@@ -572,13 +578,13 @@ def get_context_partition_and_sub_seq(self, seq_len):
...
 
-  # sub group mask generation for inference only
   def generate_masks_subgroup(self, top_k_indices, softmax_probs):
+    """subgroup mask generation for inference only"""
     # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
     batch_size, seq_len, _ = top_k_indices.shape
     cp, sub_seq = self.get_context_partition_and_sub_seq(seq_len)
 
-    #  breaking the sequence into sub sequences. It is effectively grouping the tokens in a sequence into groups, and route only within each group.
+    # break sequence into subsequences. Effectively groups tokens in a sequence into groups, & route within each group.
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/layers/moe.py:587]: Ok

File: MaxText/layers/pipeline.py [hunk 5]
```diff
@@ -546,10 +554,15 @@ def get_partition_spec_leaf(leaf):
...
-    # Outputs: Modified physical spec with "fsdp" and "fsdp_transpose" removed
-    # We may want to remove the expert sharding on attention weights as well, since
-    # those act like FSDP
+    """
+    Get physical spec without fsdp.
+
+    Inputs: original logical partition specs of all weights
+    Outputs: Modified physical spec with "fsdp" and "fsdp_transpose" removed
+
+    We may want to remove the expert sharding on attention weights as well, since those act like FSDP
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/layers/pipeline.py:563]: Ok done

File: MaxText/elastic_train.py [hunk 6]
```diff
@@ -187,7 +187,7 @@ def train_loop(config, elastic_manager, state=None):
   Args:
     config:
     state:
-    ckpt_path:
```
Comment (shralex) [inline MaxText/elastic_train.py:190]: why this change ?

File: MaxText/layers/attentions.py [hunk 7]
```diff
@@ -332,6 +336,77 @@ def generate_attention_mask(
       previous_chunk: Any = None,
       bidirectional_mask: Any = None,
   ) -> Array | None:
+    """Generates a combined attention mask for Transformer models.
```
Comment (shralex) [inline MaxText/layers/attentions.py:339]: Could you delete the comments before the def

File: MaxText/elastic_train.py (hunk 6 above)
Comment (SamuelMarks) [reply MaxText/elastic_train.py:190]: Looking at `main` you can see that the wrong name is listed in the argument list: https://github.com/AI-Hypercomputer/maxtext/blob/9cc5aec/MaxText/elastic_train.py#L190

**Review by shralex — APPROVED** [review]

File: MaxText/inference/kvcache.py [hunk 8]
```diff
@@ -141,6 +145,7 @@ def einsum_fn_with_rhs_qtensor(
     return einsum
 
   def einsum_fn_with_rhs_qtensor_and_dequant(self, value):
+    """einsum function where QTensor is the right-hand-side with special-case on float8_e4m3fn"""
```
Comment (SurbhiJainUSC) [inline MaxText/inference/kvcache.py:148]: This could be something more descriptive like:
```Get einstein summation for different dequant modes.```

File: MaxText/inference/paged_attention_kernel_v2.py [hunk 9]
```diff
@@ -79,6 +79,7 @@ def ref_ragged_paged_attention(
     sm_scale: float = 1.0,
     mask_value: float = DEFAULT_MASK_VALUE,
 ):
+  """rf ragged paged attention"""
```
Comment (SurbhiJainUSC) [inline MaxText/inference/paged_attention_kernel_v2.py:82]: what is `rf`?

File: MaxText/inference_mlperf/evaluate-accuracy.py [hunk 10]
```diff
@@ -33,7 +33,7 @@ def get_args():
 
 
 def get_groundtruth(processed_dataset_file):
-  import pandas as pd
+  import pandas as pd  # pylint: disable=import-outside-toplevel
```
Comment (SurbhiJainUSC) [inline MaxText/inference_mlperf/evaluate-accuracy.py:36]: Can we move this import to top level with other imports.

File: MaxText/inference_mlperf/trillium/select_xla_flags.py [hunk 11]
```diff
@@ -11,6 +11,9 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+"""
+Select XLA flag preconfigurations
```
Comment (SurbhiJainUSC) [inline MaxText/inference_mlperf/trillium/select_xla_flags.py:15]: nit: `XLA flag pre-configurations`

File: MaxText/input_pipeline/_hf_data_processing.py [hunk 12]
```diff
@@ -238,6 +238,7 @@ def make_hf_eval_iterator(
     global_mesh,
     process_indices_eval,
 ):
+  """make Hugging Face evaluation iterator"""
```
Comment (SurbhiJainUSC) [inline MaxText/input_pipeline/_hf_data_processing.py:241]: `Load and preprocess eval dataset, and return iterator.`

File: MaxText/input_pipeline/_input_pipeline_utils.py [hunk 13]
```diff
@@ -151,6 +151,7 @@ def __init__(
     self.unk_id = unk_id
 
   def map(self, element):
+    """map on element; returning dict of (input, targets)"""
```
Comment (SurbhiJainUSC) [inline MaxText/input_pipeline/_input_pipeline_utils.py:154]:
```
Maps a single dataset element to SFT training instance.
It concatenates the prompt and completion to form the inputs sequence.
For the targets sequence:
   - If `self.completion_only` is True, the prompt portion of the
      concatenated sequence is masked using `self.unk_id`.
   - If `self.completion_only` is False, the target sequence is 
      identical to the input sequence.
```

File: MaxText/layers/llama4.py [hunk 14]
```diff
@@ -131,9 +132,7 @@ def __call__(
     cfg = self.config
     mesh = self.mesh
 
-    assert (
-        cfg.num_experts >= 1
-    ), "Expected the Llama4 config to have num_experts > 1.  Note that MaxText only supports Llama4 Scout (MoE-only) for now!"
+    assert cfg.num_experts >= 1, "Expected the Llama4 config to have `num_experts > 1`."
```
Comment (SurbhiJainUSC) [inline MaxText/layers/llama4.py:135]: Why `Note that MaxText only supports Llama4 Scout (MoE-only) for now!` is removed?

File: MaxText/layers/pipeline.py [hunk 15]
```diff
@@ -546,10 +558,15 @@ def get_partition_spec_leaf(leaf):
...
 
   def get_physical_spec_no_fsdp(self, full_logical):
-    # Inputs: original logical partition specs of all weights
-    # Outputs: Modified physical spec with "fsdp" and "fsdp_transpose" removed
-    # We may want to remove the expert sharding on attention weights as well, since
-    # those act like FSDP
+    """
+    Get physical spec without fsdp.
+
+    Inputs: Original logical partition specs of all weights.
```
Comment (SurbhiJainUSC) [inline MaxText/layers/pipeline.py:564]: Inputs and Outputs description can be moved to `Args` section in dcostring.

File: MaxText/layers/llama4.py (hunk 14 above)
Comment (SamuelMarks) [reply MaxText/layers/llama4.py:135]: @SurbhiJainUSC It was asked to be removed by @shralex

File: MaxText/inference/kvcache.py (hunk 8 above)
Comment (SamuelMarks) [reply MaxText/inference/kvcache.py:148]: Done

File: MaxText/inference/kvcache.py [hunk 16]
```diff
@@ -104,6 +107,7 @@ def einsum_fn_with_rhs_qtensor(
       lhs_dequant_mode=None,
       lhs_calibration_mode=None,
   ):
+    """einsum function where QTensor is the right-hand-side"""
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference/kvcache.py:110]: Ok done

File: MaxText/inference_mlperf/evaluate-accuracy.py (hunk 10 above)
Comment (SamuelMarks) [reply MaxText/inference_mlperf/evaluate-accuracy.py:36]: Yes but it's a heavy import so I assume that's why it was imported here

File: MaxText/inference_mlperf/offline_mode.py [hunk 17]
```diff
@@ -219,8 +222,8 @@ def _classify_query(dataset_rows, index, query_batches):
   target_inputs = [lb[0] for lb in query_batch_keys]
   target_totals = [2 * inp for inp in target_inputs]
 
-  for i in range(len(target_inputs)):
-    if total_len <= target_totals[i] and input_len <= target_inputs[i]:
+  for i, target_input in enumerate(target_inputs):
+    if total_len <= target_totals[i] and input_len <= target_input:
       log.debug("Added sample of input length %d total_len %d for %s", input_len, total_len, query_batch_keys[i])
       return query_batch_keys[i]
   if input_len <= target_inputs[i]:
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference_mlperf/offline_mode.py:229]: What's wrong with `target_input`?

File: MaxText/input_pipeline/_hf_data_processing.py (hunk 12 above)
Comment (SamuelMarks) [reply MaxText/input_pipeline/_hf_data_processing.py:241]: Expanded

File: MaxText/input_pipeline/_input_pipeline_utils.py (hunk 13 above)
Comment (SamuelMarks) [reply MaxText/input_pipeline/_input_pipeline_utils.py:154]: Done

File: MaxText/layers/pipeline.py (hunk 15 above)
Comment (SamuelMarks) [reply MaxText/layers/pipeline.py:564]: But they aren't; that's `full_logical`. Is `full_logical` a "pytree of sharding specifications of the weights (aka self.layers.variables)"?

File: MaxText/layers/llama4.py (hunk 14 above)
Comment (shralex) [reply MaxText/layers/llama4.py:135]: We now also support Maverick.

File: MaxText/inference_mlperf/offline_mode.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply MaxText/inference_mlperf/offline_mode.py:229]: At line 229, it should be: `if input_len <= target_input:`, right?

File: MaxText/layers/pipeline.py (hunk 15 above)
Comment (SurbhiJainUSC) [reply MaxText/layers/pipeline.py:564]: `full_logical` is the original logical partition specs of all weights.
So, we can modify the docstring like this:

```
Get physical spec without fsdp.
TODO: We may want to remove the expert sharding on attention weights 
as well, since those act like fsdp.
Args:
  full_logical: original logical partition specs of all weights
Returns:
   Modified physical spec with "fsdp" and "fsdp_transpose" removed

```

File: MaxText/inference_mlperf/offline_mode.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference_mlperf/offline_mode.py:229]: No the indentation in Python dictates scope, the `i` could be initialised to `0` but `target_inputs` should never be empty so that shouldn't pose an issue.

File: MaxText/layers/pipeline.py (hunk 15 above)
Comment (SamuelMarks) [reply MaxText/layers/pipeline.py:564]: Nice. Ok: changed.

File: MaxText/inference_mlperf/offline_mode.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/inference_mlperf/offline_mode.py:229]: @mitalisi this seems like a bug -- i is not defined outside the loop. Did you mean to compare with the last element in target_inputs ? i.e., "if input_len <= target_inputs[-1]" ?

[the remark this answers is not in the document]
Comment (shralex) [reply MaxText/inference_mlperf/offline_mode.py:229]: Regardless, this bug isn't introduced in this PR, so lets not block it.

[the remark this answers is not in the document]
Comment (mitalisi) [reply MaxText/inference_mlperf/offline_mode.py:229]: Yes it is comparing with the last element, and assigning to last bucket if it does not meet the prior condition.

[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference_mlperf/offline_mode.py:229]: Technically not a bug because `i` outlives the loop; unlike other languages like C with `for(int i=0; i<…` in Python it keeps living:
```
Python 3.9.13 (v3.9.13:6de2ca5339, May 17 2022, 11:37:23) 
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> for i in range(5): pass
... 
>>> print(i)
4
````

[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference_mlperf/offline_mode.py:229]: Technically not a bug because `i` outlives the loop; unlike other languages like C with `for(int i=0; i<…` in Python it keeps living:
```
Python 3.9.13 (v3.9.13:6de2ca5339, May 17 2022, 11:37:23) 
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> for i in range(5): pass
... 
>>> print(i)
4
````

But yes `-1` would be clearer; although `i` might be more efficient

[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply MaxText/inference_mlperf/offline_mode.py:229]: @SamuelMarks - Can we change this to use `-1` for clarity? This is the only thing remaining and then we should be good to close this PR. Thanks!

[the remark this answers is not in the document]
Comment (SamuelMarks) [reply MaxText/inference_mlperf/offline_mode.py]: Ok done

File: MaxText/tests/inference/page_manager_test.py [hunk 18]
```diff
@@ -95,7 +99,7 @@ def test_initialization(self):
     self.assertTrue(jnp.all(state.page_status[1:] == 0), "All pages should be initially free (status 0)")
     self.assertTrue(jnp.all(state.num_pages_used[1:] == 0), "No pages should be used initially")
     self.assertTrue(jnp.all(state.sequence_lengths[1:] == 0), "Sequence lengths should be 0 initially")
-    self.assertTrue(jnp.all(state.has_active_page[1:] == False), "No groups should be active initially")
+    self.assertTrue(jnp.all(state.has_active_page[1:] is False), "No groups should be active initially")
```
Comment (shralex) [inline MaxText/tests/inference/page_manager_test.py:102]: There's a bug here. Before the change, the comparison was elementwise, and returned a boolean array. After the change it checks whether the entire sliced array object is False.

Comment (SamuelMarks) [reply MaxText/tests/inference/page_manager_test.py:102]: Ohh must be an incorrect lint check let me revert

**Review by SurbhiJainUSC — APPROVED** [review]
